File: src/main/java/com/example/chatService/security/CookieUtil.java
package com.example.chatService.security;

import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseCookie;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class CookieUtil {

    private static final String REFRESH_TOKEN_COOKIE = "refreshToken";
    private final JwtTokenProvider jwtTokenProvider;

    public void addRefreshTokenCookie(HttpServletResponse response, String refreshToken) {

        ResponseCookie cookie = ResponseCookie.from(REFRESH_TOKEN_COOKIE, refreshToken)
                .httpOnly(true)
                .secure(false)
                .sameSite("Lax")
                .path("/")
                .maxAge(60 * 60 * 24 * 7)
                .build();

        response.addHeader("Set-Cookie", cookie.toString());
    }

    public void clearRefreshTokenCookie(HttpServletResponse response) {

        ResponseCookie cookie = ResponseCookie.from(REFRESH_TOKEN_COOKIE, "")
                .httpOnly(true)
                .secure(false)
                .path("/")
                .maxAge(0)
                .sameSite("Lax")
                .build();

        response.addHeader("Set-Cookie", cookie.toString());
    }

    public String getRefreshToken(HttpServletRequest request) {
        Cookie[] cookies = request.getCookies();
        if (cookies == null) {
            System.out.println("No cookies in request");
            return null;
        }

        for (Cookie cookie : cookies) {
//            System.out.println("cookie = " + cookie.getName());
            if (REFRESH_TOKEN_COOKIE.equals(cookie.getName())) {
                return cookie.getValue();
            }
        }

        return null;
    }

    public String tryResolveUserFromRefreshCookie(HttpServletRequest request) {

        String refreshToken = getRefreshToken(request);
        if (refreshToken == null) return null;

        if (!jwtTokenProvider.validateToken(refreshToken)) {
            return null;
        }
        return jwtTokenProvider.getSubject(refreshToken);
    }
}

File: src/main/java/com/example/chatService/security/CookieUtil.kt
package com.example.chatService.security

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.ResponseCookie
import org.springframework.stereotype.Service

@Service
class CookieUtil(
    private val jwtTokenProvider: JwtTokenProvider
) {
    companion object {
        private const val REFRESH_TOKEN_COOKIE = "refreshToken"
    }

    fun addRefreshTokenCookie(response: HttpServletResponse, refreshToken: String) {
        val cookie = ResponseCookie.from(REFRESH_TOKEN_COOKIE, refreshToken)
            .httpOnly(true)
            .secure(false)
            .sameSite("Lax")
            .path("/")
            .maxAge((60 * 60 * 24 * 7).toLong())
            .build()
        response.addHeader("Set-Cookie", cookie.toString())
    }

    fun clearRefreshTokenCookie(response: HttpServletResponse) {
        val cookie = ResponseCookie.from(REFRESH_TOKEN_COOKIE, "")
            .httpOnly(true)
            .secure(false)
            .path("/")
            .maxAge(0)
            .sameSite("Lax")
            .build()
        response.addHeader("Set-Cookie", cookie.toString())
    }

    fun getRefreshToken(request: HttpServletRequest): String? {
        val cookies = request.cookies
        if (cookies == null) {
            println("No cookies in request")
            return null
        }
        return cookies.firstOrNull { it.name == REFRESH_TOKEN_COOKIE }?.value
    }

    fun tryResolveUserFromRefreshCookie(request: HttpServletRequest): String? {
        val refreshToken = getRefreshToken(request) ?: return null
        if (!jwtTokenProvider.validateToken(refreshToken)) return null
        return jwtTokenProvider.getSubject(refreshToken)
    }
}
